### PR TITLE
[SB-62] Creating POST and GET endpoints for transferring heartbeat data to and from DynamoDB

### PR DESCRIPTION
**Pull request by StudentCZ**
https://virgl.atlassian.net/jira/software/projects/SB/boards/1?selectedIssue=SB-86

The goal of this PR is making sure we can read and write data from and to dynamodb using our lambda endpoints. We can do a GET and POST request and have that reflect in Postman and Dynamodb.

Screen Shot 2023-01-01 at 8 56 54 PM
Screen Shot 2023-01-01 at 8 59 07 PM
Screen Shot 2023-01-01 at 8 58 22 PM

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,93 +1,71 @@
...
+# retrive the heartbeat table from DynamoDB
+dynamodb = boto3.resource(
+    "dynamodb",
+    region_name="us-east-1",
+    endpoint_url="https://dynamodb.us-east-1.amazonaws.com",
+)
+table = dynamodb.Table("heartbeat")
 
+# returns all heartbeat data in dynamodb
+@app.route("/heartbeat", methods=["GET"])
```
Comment (chunloy) [inline app.py:22]: Can we shuffle the code to organize? For example, group post routes, get routes, and function definitions together.

Let's structure the script by:
1. post routes (shorter endpoints at the top)
2. get routes (shorter endpoints at the top)
3. function definitions at the bottom of the script e.g. post_heartbeat, get_latest_heartbeats

File: app.py [hunk 2]
```diff
@@ -1,93 +1,71 @@
...
+    return jsonify(response)
 
 @app.route("/heartbeat", methods=["POST"])
 @cross_origin()
 def heartbeatpost():
     data = json.loads(request.get_data())
-    user_id = data["userId"]
-    user_role = data["userRole"]
-    time_stamp = data["timestamp"]
+    userId = data["userId"]
```
Comment (chunloy) [inline app.py:32]: Please use snake case for variable/function names. BE agreed to use snake case, let's stay consistent here. The data being stored in DynamoDB is already FE friendly e.g. camel case (see lines 53-57).

File: app.py [hunk 3]
```diff
@@ -1,93 +1,71 @@
...
+            "latitude": Decimal(str(lat)),
+            "longitude": Decimal(str(long)),
+            "speed": speed,
+        }
+    )
+
+@app.route("/heartbeat/<userId>", methods=["GET"])
+@cross_origin()
+def get_heartbeats_by_user_id(userId):
+    # convert user_id to a integer, otherwise the post request will be return a 500 error message
```
Comment (chunloy) [inline app.py:64]: It looks like the code is casting userId to a string, not an integer?

**Review by chunloy — COMMENTED** [review]
run `pip3 freeze > requirements.txt`

File: app.py [hunk 4]
```diff
@@ -1,93 +1,71 @@
...
+    "dynamodb",
+    region_name="us-east-1",
+    endpoint_url="https://dynamodb.us-east-1.amazonaws.com",
+)
+table = dynamodb.Table("heartbeat")
 
+# returns all heartbeat data in dynamodb
+@app.route("/heartbeat", methods=["GET"])
+@cross_origin()
+def get_all_heartbeat():
```
Comment (yhus) [inline app.py:24]: No limit?

File: app.py [hunk 5]
```diff
@@ -1,93 +1,71 @@
...
     return "Heartbeat data added successfully", 200
 
+def post_heartbeat(id, timestamp, lat, long, speed):
 
-# GET heartbeats by user_id
-@app.route("/heartbeat/<int:user_id>", methods=["GET"])
-def get_heartbeats_by_user_id(user_id):
-    user_exists = Heartbeat.query.filter_by(user_id=user_id).first()
-    lookback = request.args.get("lookback")
+    table.put_item(
```
Comment (yhus) [inline app.py:51]: try catch

File: app.py [hunk 6]
```diff
@@ -1,93 +1,71 @@
...
+# returns all heartbeat data in dynamodb
+@app.route("/heartbeat", methods=["GET"])
+@cross_origin()
+def get_all_heartbeat():
+    response = table.scan()["Items"]
+    return jsonify(response)
 
 @app.route("/heartbeat", methods=["POST"])
 @cross_origin()
 def heartbeatpost():
```
Comment (yhus) [inline app.py:30]: def post_heartbeat

**Review by yhus — APPROVED** [review]

File: app.py (hunk 2 above)
Comment (StudentCZ) [reply app.py]: resolved

**Review by tonyxuan2021 — APPROVED** [review]
Look like the constant are stored in the setting.yaml file now. Good to go!

**Pull request merged**